src/tracked_struct.rs

```diff
@@ -510,73 +506,61 @@ where
+            last_updated_at.is_some(),
+            "two concurrent writers to {id:?}, should not be possible"
+        );
+        if last_updated_at == Some(current_revision) {
+            // already read-locked
+            return;
+        }
 
-            // Acquire the write-lock. This can only fail if there is a parallel thread
-            // reading from this same `id`, which can only happen if the user has leaked it.
-            // Tsk tsk.
-            let swapped_out = unsafe { (*data).updated_at.swap(None) };
-            if swapped_out != last_updated_at {
-                panic!(
+        // Acquire the write-lock. This can only fail if there is a parallel thread
+        // reading from this same `id`, which can only happen if the user has leaked it.
+        // Tsk tsk.
+        let swapped_out = unsafe { (*data_raw).updated_at.swap(None) };
+        if swapped_out != last_updated_at {
+            panic!(
                 "failed to acquire write lock, id `{id:?}` must have been leaked across threads"
             );
-            }
-
-            true
-        });
-
-        if !locked {
-            return;
         }
 
-        data_raw.with_mut(|data| {
-            // UNSAFE: Marking as mut requires exclusive access for the duration of
-            // the `mut`. We have now *claimed* this data by swapping in `None`,
-            // any attempt to read concurrently will panic.
-            let data = unsafe { (*data).assume_init_mut() };
-
-            // SAFETY: We assert that the pointer to `data.revisions`
-            // is a pointer into the database referencing a value
-            // from a previous revision. As such, it continues to meet
-            // its validity invariant and any owned content also continues
-            // to meet its safety invariant.
-            unsafe {
-                if C::update_fields(
-                    current_deps.changed_at,
-                    &mut data.revisions,
-                    mem::transmute::<*mut C::Fields<'static>, *mut C::Fields<'db>>(
-                        std::ptr::addr_of_mut!(data.fields),
-                    ),
-                    fields,
-                ) {
-                    // Consider this a new tracked-struct (even though it still uses the same id)
-                    // when any non-tracked field got updated.
-                    // This should be rare and only ever happen if there's a hash collision
-                    // which makes Salsa consider two tracked structs to still be the same
-                    // even though the fields are different.
-                    // See `tracked-struct-id-field-bad-hash` for more details.
-                    data.created_at = current_revision;
-                }
-            }
-            if current_deps.durability < data.durability {
-                data.revisions = C::new_revisions(current_deps.changed_at);
+        // UNSAFE: Marking as mut requires exclusive access for the duration of
```

Comment:
I didn't review those changes too carefully because I understand that these are just reverting earlier changes?

Reply:
Yeah the changes are mostly just indentation because the `with` closure is removed.